### PR TITLE
Change answer upload to use partitioned parquet files [#173607905]

### PR DESCRIPTION
**Pull request by dougmartin**

### CODE REVIEW

**Comment by dougmartin** [comment]
@scytacki this is a draft request.  I'm waiting for the composite query to finish and then I can change the `orderBy` and remove the test code I have that stops after 10 batches.

**Comment by dougmartin** [comment]
@scytacki wait on this - it has been running for a while and the memory usage is growing.  I changed the callback to process more to use a setTimeout to avoid blowing up the stack but it is consuming a lot of memory.

**Comment by dougmartin** [comment]
@scytacki ok, this is good to review.  I stopped the upload and remove the setTimeout and the memory usage is stable now.

**Comment by scytacki** [comment]
I added a section to the doc about partitioning: https://docs.google.com/document/d/1F4ozfQOjzZfMMYdGiPp0MGWpAxjTbMa6lDIvPSnuNcI/edit#heading=h.8zbmxyy6np5y

In that section I also documented the usage of bucketing. However as far as I can tell it won't benefit us right now.

File: scripts/export-answers.js [hunk 1]
```diff
@@ -23,57 +34,154 @@ if (!fs.existsSync(process.env.GOOGLE_APPLICATION_CREDENTIALS)) {
...
+  }
+
+  const answers = answerHash[runKey];
+  const answer = answers[0];
+  const filename = path.join(outputPath, `${answer.run_key}.parquet`);
+  const folder = answer.resource_url.replace(/[^a-z0-9]/g, "-");
+  const key = `${DIRECTORY}/${folder}/${runKey}.parquet.gz`;
+
+  if (uploadPromises[runKey]) {
+    console.error("Already uploading", runKey)
```
[the remark this answers is not in the document]
Comment (dougmartin) [reply scripts/export-answers.js:94]: This is really just here for a sanity check.  When running the script I redirect stderr to a log file and this error log didn't get tripped for any of the uploads.

File: scripts/export-answers.js [hunk 2]
```diff
@@ -23,57 +34,154 @@ if (!fs.existsSync(process.env.GOOGLE_APPLICATION_CREDENTIALS)) {
...
+    } catch (err) {
+      console.error(err);
+    } finally {
+      await deleteFile();
+    }
+
+    // while running keep the upload promises object as small as possible
+    if (!waitingForUploadsToFinish) {
+      delete uploadPromises[runKey];
+    }
```
[the remark this answers is not in the document]
Comment (dougmartin) [reply scripts/export-answers.js:141]: I've added some comments in the code to hopefully make it more clear.

File: scripts/export-answers.js [hunk 3]
```diff
@@ -23,57 +34,154 @@ if (!fs.existsSync(process.env.GOOGLE_APPLICATION_CREDENTIALS)) {
...
-    }, function (err) {
-      if (err) {
-        console.error(`${documentSnapshot.id}: ${err.toString()}`)
-      }
-    });
+    const answer = documentSnapshot.data();
+    answerHash[answer.run_key] = answerHash[answer.run_key] || [];
+    answerHash[answer.run_key].push(answer);
+    if (lastRunKey && (answer.run_key !== lastRunKey)) {
+      uploadAnswers(lastRunKey)
```
[the remark this answers is not in the document]
Comment (dougmartin) [reply scripts/export-answers.js:154]: I've removed this - it was vestigial code from an earlier way I was tracking the uploads.

**Pull request merged**